### ISSUE TITLE
Limitation of Polyfit Estimate - 34

### ISSUE THREAD

**Issue by mssamhan31** [mssamhan31/Solar-Curtailment-CANVAS#3]
When I run the VVAr Script using Polyfit Estimate made by Tim to estimate the expected generation without curtailment, I expected obtaining result higher than the actual energy. However, it is sometime lower, so the curtailment is negative.

Needs to improve the polyfit estimate. Some ideas:
- Never make the actual value lower than the fitting?
- Define the peak time?

**Comment by mssamhan31** [comment]
picture:

![image](https://user-images.githubusercontent.com/110155265/188060627-711b68e0-cebd-45aa-b21c-03840049adaf.png)

**Comment by mssamhan31** [comment]
WIP ....

**Comment by mssamhan31** [comment]
![image](https://user-images.githubusercontent.com/110155265/188336459-3d556aee-8dba-4ab0-accd-dd36229867fb.png)
With this example, we can say that it is possible that the maximum real power is not the same with the maximum expected real power

**Comment by mssamhan31** [comment]
![image](https://user-images.githubusercontent.com/110155265/188336545-2333a6f3-d298-47cb-b817-6823741a3db9.png)
By this example, we should be careful with the decreasing gradient filter by Tim. In this case, if we apply the decreasing gradient filter by Tim, we will pass the points in the morning (which seems like curtailed by cloud), and filter out the points in the early afternoon (which suddenly has higher gradient than the morning).. consider delete Tim's decerasing gradient filter?

**Comment by mssamhan31** [comment]
![image](https://user-images.githubusercontent.com/110155265/188336662-c59dd637-d410-4b9a-910c-f1e31918277b.png)

This is a sample for tripping case in a non clear sky day, which Naomi believes that we should not use a polyfit estimate because of too much distraction. I think, Tim's first filter (only includes increasing point until the peak) is genius. If we apply Tim's first filter and add more constrain to the polyfit, we should consider the probability of implementing polyfit.

**Comment by mssamhan31** [comment]
So, the approach for optimizing the polyfit estimate might be:
1. Remove Tim's decreasing gradient filter (which makes issue in cloudy morning)
2. Make the polyfit estimate NEVER BELOW the actual power (it is nonsense!)
3. Make sure the first x-intercept of the polyfit never higher than the first x-intercept of the real power curve. It is possible, however, that the first x-intercept of the polyfit to be lower than the first x-intercept of the real power curve, when it is cloudy on sunrise.
4. Make sure the second x-intercept of the polyfit never lower than the second x-intercept of the real power curve. It is possible, however, that the second x-intercept of the polyfit to be higher than the second x-intercept of the real power curve, when it is cloudy on sunset.

Let's research how to implement it in python!

**Comment by mssamhan31** [comment]
Also, some ideas:
1. possible first x-intercept must be between surise time at summer and winter (5.30 - 7.00). Probably the most useful bound is the 7.00 bound. When the polyfit result gives first x-intercept bigger than 7.00 it must be because of cloud.
2. possible second x-intercept must be between sunset time at winter and summer (16.45 - 20.15). Probably the most useful bound is the 16.45 bound. When the polyfit reslt gives second x-intercept lower than 16.45 it must be because of cloud.

**Comment by mssamhan31** [comment]
References:
https://realpython.com/python-scipy-cluster-optimize/#using-the-optimize-module-in-scipy
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.NonlinearConstraint.html#scipy.optimize.NonlinearConstraint
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html

**Comment by mssamhan31** [comment]
Use scipy.optimize.minimize with 'trust-constr' method to add constrain of power expected never lower than real power. Result is better!

sample 1:
![image](https://user-images.githubusercontent.com/110155265/188380911-e7286bcc-3953-4620-acdc-df30a5327257.png)

sample 2:
![image](https://user-images.githubusercontent.com/110155265/188380994-117f3e57-5f42-411f-8c79-7fe428956666.png)

sample 3:
![image](https://user-images.githubusercontent.com/110155265/188380815-f6a3a803-ced2-41e3-9c9a-31e68e716d7b.png)

sample 4:
![image](https://user-images.githubusercontent.com/110155265/188516717-435fcf9a-9772-48f9-afa8-ab1f36520d35.png)
next will try to limit the x-intercept value to increase accuracy for sample 4 (which the first x-intercept seems to be lower than actual value)

**Comment by mssamhan31** [comment]
also, sample 4 seems to be weird as well... but still no idea what to do

**Comment by mssamhan31** [comment]
to ponder: could we actually infer all the remaining curtailment MUST BE because of cloud? For example, in sample 3, in the afternoon there are vwatt curtailment. However, in the morning and in the late afternoon, the power generated is lower because of cloud.

**Comment by mssamhan31** [comment]
Added the bound, the polyfit must intersect x axis at sunrise and sunset, which is detected by the first time and last time the power expected is >0. Result:

Sample 1:
![image](https://user-images.githubusercontent.com/110155265/188518172-d5ae08bf-863e-4619-b889-f258715a0b0c.png)

Sample 2:
![image](https://user-images.githubusercontent.com/110155265/188519343-9ee20e3b-9921-4f04-bd7f-9176a4216abe.png)

Analysis:
The result is nice for sample 1, but not for sample 2. The reason might be in sample 2, the power increases rapidly in the morning so the gradient is really high, and the polyfit's estimation will be much higher than the actual expected power.

I have also tried to add the 4th constraint, which is the upper bound, making sure that the expected power will never be higher than the actual ac value. However, the running time is more than 5 minutes and it finally didn't get the solution (probably because there is no solution).

So, let's just decide to keep using only this following process:
1. Filter data only from sunrise and sunset (first and last non zero power value)
2. Filter sudden curtailment (power data which is lower than the previous and next data point)
3. Use least squares method to find the best quadratic polyfit, with constraint: the expected power never below the actual power.
